### PR TITLE
Events for HorizontalPodAutoscalers.

### PR DESCRIPTION
**Pull request by jszczepkowski**
Implemented events for reconcilation of HorizontalPodAutoscalers.

### CODE REVIEW

**Comment by jszczepkowski** [comment]
Part of #12087.

File: pkg/controller/podautoscaler/horizontal.go [hunk 1]
```diff
@@ -43,15 +44,21 @@ const (
...
+	eventRecorder record.EventRecorder
 }
 
 var downscaleForbiddenWindow, _ = time.ParseDuration("20m")
 var upscaleForbiddenWindow, _ = time.ParseDuration("3m")
 
 func NewHorizontalController(client client.Interface, metricsClient metrics.MetricsClient) *HorizontalController {
+	broadcaster := record.NewBroadcaster()
+	broadcaster.StartRecordingToSink(client.Events(""))
+	recorder := broadcaster.NewRecorder(api.EventSource{Component: "horizontal-controller"})
```
Comment (mwielgus) [inline pkg/controller/podautoscaler/horizontal.go]: Maybe include autoscaling in the name.

File: pkg/controller/podautoscaler/horizontal.go [hunk 2]
```diff
@@ -63,92 +70,101 @@ func (a *HorizontalController) Run(syncPeriod time.Duration) {
...
-func (a *HorizontalController) reconcileAutoscalers() error {
-	ns := api.NamespaceAll
-	list, err := a.client.Experimental().HorizontalPodAutoscalers(ns).List(labels.Everything(), fields.Everything())
+func (a *HorizontalController) reconcileAutoscaler(hpa experimental.HorizontalPodAutoscaler) error {
+	reference := fmt.Sprintf("%s/%s/%s", hpa.Spec.ScaleRef.Kind, hpa.Spec.ScaleRef.Namespace, hpa.Spec.ScaleRef.Name)
+
+	scale, err := a.client.Experimental().Scales(hpa.Spec.ScaleRef.Namespace).Get(hpa.Spec.ScaleRef.Kind, hpa.Spec.ScaleRef.Name)
 	if err != nil {
-		return fmt.Errorf("error listing nodes: %v", err)
+		a.eventRecorder.Event(&hpa, "failed to query scale subresource", err.Error())
```
Comment (mwielgus) [inline pkg/controller/podautoscaler/horizontal.go]: Just to confirm - is starting from lowercase a convention for events as well?

File: pkg/controller/podautoscaler/horizontal.go (hunk 1 above)
Comment (jszczepkowski) [inline pkg/controller/podautoscaler/horizontal.go]: Done.

File: pkg/controller/podautoscaler/horizontal.go (hunk 2 above)
Comment (jszczepkowski) [inline pkg/controller/podautoscaler/horizontal.go]: Fixed.

I was basing on events generated by service controller. But after careful investigation, our convention is different: https://github.com/kubernetes/kubernetes/blob/master/docs/devel/api-conventions.md#events, and events generated by service controller are not correct. Thanks for noticing this!

**Comment by jszczepkowski** [comment]
@mwielgus Comments applied, PTAL

**Comment by mwielgus** [comment]
LGTM

File: pkg/controller/podautoscaler/horizontal.go (hunk 2 above)
Comment (jszczepkowski) [inline pkg/controller/podautoscaler/horizontal.go]: I've prepared a follow-up: https://github.com/kubernetes/kubernetes/pull/13922

**Pull request merged**